activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb

```diff
@@ -55,12 +55,15 @@ def to_sql_and_binds(arel_or_sql_string, binds = [], preparable = nil, allow_ret
       # can be used to query the database repeatedly.
       def cacheable_query(klass, arel) # :nodoc:
         if prepared_statements
+          collector = collector()
+          collector.retryable = true
```

Comment:
What makes me uneasy here is that we assume all `cacheable_query` are retryiable. That may be true today, but might not be forever. Perhaps one day someone will use `cacheable_query` for `INSERT` or something like that.

Reply:
This is the same approach taken in `to_sql_and_binds` above. We start by setting `retryable: true` on the `collector`, and then the `ToSql` visitor will change it to `false` if it encounters an `Arel::Node` that would make the query "not retryable". If the query includes an `INSERT`, I would expect the `ToSql` visitor to handle that case already.